Create brainsuite.py

jason-wg opened the pull request:
I created an interface for Brainsuite's command line tools. I'd like it to be included in nipype builds installed per the instructions on the nipype website.

satra commented:
it would be good to run a PEP8 checker on this code. also in many cases you can use the simpler form for instantiating names of outputs using `name_source` and `name_template` as shown here:

http://nipy.org/nipype/devel/cmd_interface_devel.html#creating-outputs-on-the-fly

jason-wg commented:
I updated the formatting to comply with the PEP8 style guide (The exception is that some lines run over 79 characters)

As for instantiating names, the reason I approached it this way is to shorten the names of outputs that ran through multiple steps of processing using brainsuite command line tools. Is it fine to leave the instantiation of outputs how it is now?

satra commented on nipype/interfaces/brainsuite.py:
```diff
@@ -0,0 +1,980 @@
+        else:
+            outputs['outputMRIVolume'] = self._gen_filename('outputMRIVolume')
+
+        if isdefined(self.inputs.outputMaskFile):
```
these almost seem like a pattern and i would like to simplify their creation. we can work on this after next week.

jason-wg commented:
@satra 
Can you please take a look at these updates?
I made a helper function that simplifies the instantiation of output names/reduces repitition in the code.

chrisfilo commented on nipype/interfaces/brainsuite.py:
```diff
@@ -0,0 +1,719 @@
+)
+
+
+class bseInputSpec(CommandLineInputSpec):
```
Classes in the rest of Nipype codebase begin with a capitol letter...

chrisfilo commented on nipype/interfaces/brainsuite.py:
```diff
@@ -0,0 +1,718 @@
+    _cmd = 'bse'
+
+    def _gen_filename(self, name):
+        exec('is_user_defined = isdefined(self.inputs.' + name + ')')
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use of exec](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2165532840/issues/source?bid=2767207&fileBranchId=3034582#l73)

chrisfilo commented on nipype/interfaces/brainsuite.py:
```diff
@@ -0,0 +1,718 @@
+        exec('is_user_defined = isdefined(self.inputs.' + name + ')')
+
+        if is_user_defined:
+            exec("toReturn = os.path.abspath(self.inputs." + name + ")")
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use of exec](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2165532840/issues/source?bid=2767207&fileBranchId=3034582#l76)

jason-wg commented:
Thanks, I removed the use of exec, and further simplified _gen_filename

satra commented on nipype/interfaces/brainsuite.py:
```diff
@@ -0,0 +1,675 @@
+
+    inputMRIFile = File(exists=True, mandatory=True, argstr='-i %s',
+                        desc='input MRI volume', position=0)
+    outputMRIVolume = File(mandatory=False,
```
the `mandatory=False` can all be removed.

jason-wg commented:
Removed unnecessary metadata

satra commented:
the next few steps are:
- each class needs a description and an example (check the spm/fsl interfaces)
- `make specs`
- `make check-before-commit`

it would also be useful for future users to add an example for brainsuite, and add it to the examples folder, and test it using run_tests. make sure to modify .circle.yml to add installing brainsuite.

before running `make specs`, make sure it's merged with master. then use `git add` + `commit` to add these autogenerated tests for your interfaces.

before you do the above, you should decide if you want to isolate brainsuite in its own folder. this may be a reasonable thing if you are planning to dissociate components of brainsuite (e.g., see the spm, fsl, freesurfer, ants, interfaces)

jason-wg commented:
@satra 
I finished adding description and example to each class, and I ran the doctests manually too.

I ran make specs and make check-before-commit.
In 9c12e44 I added the tests that were autogenerated for my classes (I was logged onto my other account, jwongCode)

Running make specs and then make check-before-commit also changed/created some other files that seemed to be unrelated to my interface, and I added them in 515c16e

_note: make check-before-commit resulted in some failures in a few doctests contained in files that I hadn't edited, for example: could not locate the file called ./xxx that is referenced in https://github.com/jason-wg/nipype/blob/patch-1/nipype/workflows/dmri/fsl/tbss.py
They were from fsl tests, so could this be from my installation of fsl?_

Finally, yes, it's fine to keep Brainsuite in its own folder.

satra commented:
@jason-wg - thanks - i see some changes for s3 in io.py. were these things you made? can you perhaps do a diff with an separate download of the master nipype to make sure. every once in a while github does some weird compares.

jason-wg commented:
@satra 
No, I didnt actually make any changes in s3 in io.py.
I just noticed those changes too, and a diff with a separate download from master nipype does show differences. 
I remember a few months ago I had in fact made changes to io.py to point out an issue while looking at #1316 
Can we ignore the changes to io.py in my branch?

satra commented:
unfortunately github won't allow that through an automated merge. could you simply copy over the io.py and commit it and regenerate the tests or revert the datasink tests that were created.

jason-wg commented:
@satra 
I copied over io.py from nipype master, and also reverted datasink tests to match nipype master, and deleted the additional S3DataSink that was generated from my old io.py

satra commented:
thanks -  a few more things. 

- appears to have one more data sink related file
- update the CHANGES file

jason-wg commented:
@satra 
I udated the CHANGES file, and removed the additional DataSink related files

satra commented:
thanks @jason-wg - will wait for all the test to be done.

jason-wg commented:
@satra 
Sounds good. Thanks for all your feedback on this PR
